test/libevmasm/Assembler.cpp

```diff
@@ -229,8 +229,8 @@ BOOST_AUTO_TEST_CASE(immutables_and_its_source_maps)
 			auto const NumExpectedMappings =
 				(
 					2 +                        // PUSH <a> PUSH <b>
-					(numActualRefs - 1) * 5 +  // DUP DUP PUSH <n> ADD MTOSRE
-					3                          // PUSH <n> ADD MSTORkhbE
+					(numActualRefs - 1) * 5 +  // DUP DUP PUSH <n> ADD MSTORE
+					3                          // PUSH <n> ADD MSTORE
```

Comment:
@christianparpart , what are the assumption on the stack before the `DUP` opcodes? What are we dupping and is `<n>` supposed to be the immutable value?

this was introduced in:
- https://github.com/ethereum/solidity/commit/b1dd0d0d02c96083ca8550f67dab7b54281a42e7

Reply:
Based on the [`BOOST_AUTO_TEST_CASE(immutable)`](https://github.com/ethereum/solidity/blob/d2f86ffb16fbdc95a4cab0fae8273701ac777785/test/libevmasm/Assembler.cpp#L302) it seems like on top of the stack we have:
```
<CODE_OFFSET> <IMMUTABLE_VALUE>
```

and `<n>`. represents the relative offset of the immutable position to the code offset.

Perhaps the assumptions about the stack should be documented and wether we are always using `DUP2 DUP2`.

Reply:
snippet from [`Assembly::assemble()`](https://github.com/ethereum/solidity/blob/d2f86ffb16fbdc95a4cab0fae8273701ac777785/libevmasm/Assembly.cpp#L630-L655):

```cpp
case AssignImmutable:
{
	// Expect 2 elements on stack (source, dest_base)
	auto const& offsets = immutableReferencesBySub[i.data()].second;
	for (size_t i = 0; i < offsets.size(); ++i)
	{
		if (i != offsets.size() - 1)
		{
			ret.bytecode.push_back(uint8_t(Instruction::DUP2));
			ret.bytecode.push_back(uint8_t(Instruction::DUP2));
		}
		// TODO: should we make use of the constant optimizer methods for pushing the offsets?
		bytes offsetBytes = toCompactBigEndian(u256(offsets[i]));
		ret.bytecode.push_back(static_cast<uint8_t>(pushInstruction(static_cast<unsigned>(offsetBytes.size()))));
		ret.bytecode += offsetBytes;
		ret.bytecode.push_back(uint8_t(Instruction::ADD));
		ret.bytecode.push_back(uint8_t(Instruction::MSTORE));
	}
	if (offsets.empty())
	{
		ret.bytecode.push_back(uint8_t(Instruction::POP));
		ret.bytecode.push_back(uint8_t(Instruction::POP));
	}
	immutableReferencesBySub.erase(i.data());
	break;
}
```

Reply:
> Perhaps the assumptions about the stack should be documented and wether we are always using `DUP2 DUP2`.

Maybe that's a good idea, but then it would be better if done in other PR.